src/internal-links/suggestions-generator.js

```diff
@@ -0,0 +1,151 @@
+
+  const processBatch = async (batch, urlTo) => {
+    try {
+      const requestBody = await getPrompt({ alternative_urls: batch, broken_url: urlTo }, 'broken-backlinks', log);
+      await sleep(1000);
+      const response = await firefallClient.fetchChatCompletion(requestBody, firefallOptions);
+
+      if (response.choices?.length >= 1 && response.choices[0].finish_reason !== 'stop') {
+        log.error(`No suggestions found for ${urlTo}`);
+        return null;
+      }
+
+      return JSON.parse(response.choices[0].message.content);
+    } catch (error) {
+      log.error(`Batch processing error: ${error.message}`);
+      return null;
+    }
+  };
+
+  const processLink = async (link, headerSuggestions) => {
+    log.info(`Processing link: ${link.url_to}`);
+    const suggestions = [];
+    for (const batch of dataBatches) {
+      // eslint-disable-next-line no-await-in-loop
+      const result = await processBatch(batch, link.url_to);
+      if (result) {
+        suggestions.push(result);
+      }
+    }
+
+    if (totalBatches > 1) {
+      log.info(`Compiling final suggestions for: ${link.url_to}`);
+      try {
+        const finalRequestBody = await getPrompt({ suggested_urls: suggestions, header_links: headerSuggestions, broken_url: link.url_to }, 'broken-backlinks-followup', log);
+        await sleep(1000);
+        const finalResponse = await firefallClient
+          .fetchChatCompletion(finalRequestBody, firefallOptions);
+
+        if (finalResponse.choices?.length >= 1 && finalResponse.choices[0].finish_reason !== 'stop') {
+          log.error(`No final suggestions found for ${link.url_to}`);
+          return { ...link };
+        }
+
+        const answer = JSON.parse(finalResponse.choices[0].message.content);
+        log.info(`Final suggestion for ${link.url_to}: ${JSON.stringify(answer)}`);
+        return {
+          ...link,
+          urls_suggested: answer.suggested_urls?.length > 0 ? answer.suggested_urls : [finalUrl],
+          ai_rationale: answer.ai_rationale?.length > 0 ? answer.ai_rationale : 'No suitable suggestions found',
+        };
+      } catch (error) {
+        log.error(`Final suggestion error for ${link.url_to}: ${error.message}`);
+        return { ...link };
+      }
+    }
+
+    log.info(`Suggestions for ${link.url_to}: ${JSON.stringify(suggestions[0]?.suggested_urls)}`);
+    return {
+      ...link,
+      urls_suggested:
+        suggestions[0]?.suggested_urls?.length > 0 ? suggestions[0]?.suggested_urls : [finalUrl],
+      ai_rationale:
+        suggestions[0]?.ai_rationale?.length > 0 ? suggestions[0]?.ai_rationale : 'No suitable suggestions found',
+    };
+  };
+
+  const headerSuggestionsResults = [];
+  for (const link of auditData.auditResult.brokenInternalLinks) {
+    try {
+      // eslint-disable-next-line no-await-in-loop
+      const requestBody = await getPrompt({ alternative_urls: headerLinks, broken_url: link.url_to }, 'broken-backlinks', log);
```

Comment:
Does it make sense use Promise.all()?

This also raises a question about how we handle errors within the loop.
Should we explicitly return null in case of an error? → headerSuggestionsResults.push(null);
Do we add a suggestion in cases of partial success?

Reply:
I would like to use promise.all but promise.all fails even if a single promise in the list fails which we do not want. 
Inserting null in headerSuggestionResults array is important because we are using this array and location of each result in later processing as `headerSuggestionsResults[index]`.